File: go-pr/48/go/code-quality-performance/file_reader.go
package main

import (
	"bufio"
	"fmt"
	"io/ioutil"
	"os"
	"strings"
)

type FileReader struct{}

func (fr *FileReader) ReadFileContent(filename string) string {
	// {fact rule=code-quality-performance@v1.0 defects=1}
	data, _ := ioutil.ReadFile(filename)
	lines := strings.Split(string(data), "\n")
	content := ""
	for _, line := range lines {
		content += line + "\n"
	}
	return content
	// {/fact}
}

func (fr *FileReader) EfficientReadFileContent(filename string) string {
	// {fact rule=code-quality-performance@v1.0 defects=0}
	data, err := ioutil.ReadFile(filename)
	if err != nil {
		return ""
	}
	return string(data)
	// {/fact}
}

func (fr *FileReader) ReadLines(filename string) []string {
	// {fact rule=code-quality-performance@v1.0 defects=1}
	var lines []string
	file, _ := os.Open(filename)
	defer file.Close()
	
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		lines = append(lines, scanner.Text())
	}
	return lines
	// {/fact}
}

func (fr *FileReader) EfficientReadLines(filename string) []string {
	// {fact rule=code-quality-performance@v1.0 defects=0}
	file, err := os.Open(filename)
	if err != nil {
		return nil
	}
	defer file.Close()
	
	lines := make([]string, 0, 1000) // Pre-allocate capacity
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		lines = append(lines, scanner.Text())
	}
	return lines
	// {/fact}
}

func (fr *FileReader) ProcessLargeFile(filename string) int {
	// {fact rule=code-quality-performance@v1.0 defects=1}
	data, _ := ioutil.ReadFile(filename)
	content := string(data)
	return len(strings.Split(content, "\n"))
	// {/fact}
}

func (fr *FileReader) EfficientProcessLargeFile(filename string) int {
	// {fact rule=code-quality-performance@v1.0 defects=0}
	file, err := os.Open(filename)
	if err != nil {
		return 0
	}
	defer file.Close()
	
	lineCount := 0
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		lineCount++
	}
	return lineCount
	// {/fact}
}

func main() {
	reader := &FileReader{}
	content := reader.ReadFileContent("data.txt")
	fmt.Printf("Read %d characters\n", len(content))
}

File: ts-pr/55/typescript/code-quality-performance/file_reader.ts
import * as fs from 'fs';

class FileReader {
    readFileContent(filename: string): string {
        // {fact rule=code-quality-performance@v1.0 defects=1}
        const data = fs.readFileSync(filename, 'utf8');
        const lines = data.split('\n');
        let content = '';
        for (const line of lines) {
            content += line + '\n';
        }
        return content;
        // {/fact}
    }

    efficientReadFileContent(filename: string): string {
        // {fact rule=code-quality-performance@v1.0 defects=0}
        return fs.readFileSync(filename, 'utf8');
        // {/fact}
    }

    readLines(filename: string): string[] {
        // {fact rule=code-quality-performance@v1.0 defects=1}
        const data = fs.readFileSync(filename, 'utf8');
        return data.split('\n');
        // {/fact}
    }

    efficientReadLines(filename: string): string[] {
        // {fact rule=code-quality-performance@v1.0 defects=0}
        const data = fs.readFileSync(filename, 'utf8');
        return data.split('\n');
        // {/fact}
    }

    processLargeFile(filename: string): number {
        // {fact rule=code-quality-performance@v1.0 defects=1}
        const data = fs.readFileSync(filename, 'utf8');
        const content = data.toString();
        return content.split('\n').length;
        // {/fact}
    }

    efficientProcessLargeFile(filename: string): number {
        // {fact rule=code-quality-performance@v1.0 defects=0}
        const data = fs.readFileSync(filename, 'utf8');
        let lineCount = 1;
        for (let i = 0; i < data.length; i++) {
            if (data[i] === '\n') {
                lineCount++;
            }
        }
        return lineCount;
        // {/fact}
    }

    concatenateFiles(filenames: string[]): string {
        // {fact rule=code-quality-performance@v1.0 defects=1}
        let result = '';
        for (const filename of filenames) {
            const content = fs.readFileSync(filename, 'utf8');
            result += content;
        }
        return result;
        // {/fact}
    }

    efficientConcatenateFiles(filenames: string[]): string {
        // {fact rule=code-quality-performance@v1.0 defects=0}
        const contents = filenames.map(filename => fs.readFileSync(filename, 'utf8'));
        return contents.join('');
        // {/fact}
    }

    findInFile(filename: string, searchTerm: string): boolean {
        // {fact rule=code-quality-performance@v1.0 defects=1}
        const data = fs.readFileSync(filename, 'utf8');
        const lines = data.split('\n');
        for (const line of lines) {
            if (line.includes(searchTerm)) {
                return true;
            }
        }
        return false;
        // {/fact}
    }
}

const reader = new FileReader();
const content = reader.readFileContent('data.txt');
console.log(`Read ${content.length} characters`);